Update aws-sdk to v2

fusion2004 opened the pull request:
Heya, this is a first attempt at upgrading the gem to use v2 of the aws-sdk gem.

The entire test suite passes, minus one test I've marked as pending. The new version of aws-sdk requires a region to be specified at instantiation of resources and clients, but we currently have a test specifically checking that you can instantiate the connection without **any** credentials. I'd recommend we just delete the test.

Here are some notes:
* `:public_read` is no longer a valid ACL, I've replaced instances in tests with what is valid: `'public-read'`
* I had to introduce a `CarrierWave::SanitizedFile` instance in the `AWSFile` spec since the implementation now relies on the `to_file` defined in it. If needed/wanted, I can keep fiddling with the test doubles and perhaps get this to work without it.
* I'm not sure what is the earliest version of v2 we should support, so I've specified the latest. Once we settle on the implementation, maybe we can just keep rolling back the version until we find a version that doesn't work with what we've written and call the version directly after that the earliest supported?
* I'm not certain how the numbers in the readme for Disk Space, Lines of Code, Boot Time, Runtime Deps, and Develop Deps were generated, so I've left them alone.

sorentwo commented:
That was quick! Thanks for the PR.

All of the changes look good and mostly minor.  From my review I see at least two breaking changes here:

* region is required
* acl must be specified as a string (is that correct?)

Those are pretty trivial breaks though, and this is pre 1.0, so a minor version bump will suffice.

Honestly I don't recall why there was a spec about instantiating without any credentials. There is a required attribute now though, so it can't stay. Go ahead and delete it.

The readme needs to be updated as well as the change log. If you can make those changes I'll gladly merge this in.

ryanjohns commented:
A potential reason for having a spec with no credentials is because credentials are not required when using the aws-sdk from an EC2 instance with an IAM role. In this scenario the EC2 metadata service is used to fetch a dynamically assigned set of credentials. I don't actually use it this way in my project so I can't really comment any more about it and I'm not sure how to test this behavior from outside of EC2 anyway.

fusion2004 commented:
Technically, the ACL does not need to be a string, it's just that it uses dashes instead of underscores and this is not valid: `:public-read`.

If we want to keep using and recommending a symbol, we can use `:'public-read'`

fusion2004 commented:
So looking through, I realized I missed fixing getting the URL in my first pass, so I fixed that and associated tests, and added a feature test for getting the public URL so that doesn't happen again.

I also switched to using a symbol again for the ACL everywhere, but it shouldn't matter whether users of the gem use a symbol or string.

fusion2004 commented:
Okay, I've updated the changelog and readme.

I removed the section about using a custom `AWS.config` as that doesn't exist anymore. Custom configuration options can just be specified as part of `config.aws_credentials`.

sorentwo commented:
@ryanjohns: That's a use case that I hadn't even thought of, and likewise isn't done at my company. The test doesn't provide much in the way of documentation, and isn't possible to test continuously in an EC2 environment, so we'll just let it die. Thanks for your commentary, it enriched the discussion =)

@fusion2004: I'm not sure if you added the additional notes to the lead comment after opening the PR or if I just missed them, but I'll address them all now:

* Nice documentation around `public_read` vs `public-read`. I don't think the forced symbolization is really necessary, but it's a single declaration, it certainly won't hurt.
* Since my initial development on this gem I've used stubs less and less. At this point I truly favor real objects, so the use of `SanitizedFile` is a welcome change.
* Supporting the latest release of V2 makes the most sense. If any requests come up for looser version constraints we can look at rolling back the numbers.
* At this point I don't exactly recally how the Fog vs AWS numbers were generated. They have a note about the time at which they were relevant, I think it still gets the point across.

Thanks for all of your work on this. I'll try upgrading and running this through our staging environment. If everything works as expected I'll cut `0.6.0`.

Soliah commented:
The discussion around removing the test without credentials doesn't affect the API right? We currently use IAM Roles and don't want to have to pass the config options.

sorentwo commented:
@Soliah: No, the API isn't effected at all. The spec was simply removed because it didn't apply anymore. All that changed was the region requirement.

forthowin commented:
I've had problem using `public-read` and switched to using `public_read` and the url generated when i retrieve an image from S3 is working as expected.
